components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java

```diff
@@ -709,26 +687,104 @@ private boolean isDomainNameEmbeddedInAttributeValue(String filterOperation, Str
+                int numberOfFilteredUsers = userNames.length;
+                if (numberOfFilteredUsers <= 0 && offset > 1) {
+                    offset = calculateOffset(condition, offset, sortBy, sortOrder, domainName);
+                } else {
+                    // Returned user names size > 0 implies there are users in that domain which is larger than the offset.
+                    offset = 1;
+                    limit = calculateLimit(limit, numberOfFilteredUsers);
+                }
+                filteredUsernames.addAll(Arrays.asList(userNames));
+
+                // If the limit is changed then filtering needs to be stopped.
+                if (limit == 0)
+                    break;
             }
             return filteredUsernames.toArray(new String[0]);
         }
     }
 
+    /**
+     * Method to update the count(limit) when iterating a filter across all domains.
+     *
+     * @param limit                 Counting value (limit)
+     * @param numberOfFilteredUsers Amount of users filtered in the search
+     * @return Calculated new limit (count)
+     */
+    private int calculateLimit(int limit, int numberOfFilteredUsers) {
+        if (log.isDebugEnabled()) {
+            log.debug("Updating the original count to be used for the next domain search. ");
+        }
+        limit = limit - numberOfFilteredUsers;
+        if (limit < 0) {
+            limit = 0;
+        }
+        return limit;
+    }
+
+    /**
+     * Method to update the offset when iterating a filter across all domains.
+     *
+     * @param condition  Condition of the single attribute filter
+     * @param offset     Starting index
+     * @param sortBy     Sort by
+     * @param sortOrder  Sort order
+     * @param domainName Domain to be filtered
+     * @return New calculated offset
+     * @throws CharonException Error while filtering the domain from index 1 to offset
+     */
+    private int calculateOffset(Condition condition, int offset, String sortBy, String sortOrder, String domainName)
+            throws CharonException {
+        if (log.isDebugEnabled()) {
+            log.debug(String.format("Filter returned no results for original offset: %d. Checking for number of "
+                    + "matches from the beginning to the original offset: %d for the same filter and updating "
+                    + "the new offset", offset, offset));
+        }
+        // Starting index of the filter
+        int initialOffset = 1;
+
+        // Checking the number of matches till the original offset.
+        int skippedUserCount;
+        String[] skippedUsers = getFilteredUsernames(condition, initialOffset, offset, sortBy, sortOrder, domainName);
+        if (skippedUsers == null) {
+            skippedUserCount = 0;
+        } else {
+            skippedUserCount = skippedUsers.length;
+        }
+
+        // Calculate the new offset and return
+        return offset - skippedUserCount;
```

Comment:
What would be returned for the `skippedUsers == null` scenario?

Reply:
In line 777, when the skippedUsers == null, skippedUserCount=0

Reply:
@somindatommy skippedUsers is null only when there are no users in the domain?

Reply:
Actually my understanding is that it never returns any null value. But i was asked to do it as a precaution.